# Where to put a 99-Ranch or H-Mart in San Antonio?

by Andrew Metzger

San Antonio, despite being the 7th most populous city in the country and the 2nd most populous city in the state of Texas, has no "China town". According to the U.S. Census, San Antonio also has a similar Asian population to Dallas (45k vs 50k) which has its own Asian food district. Even Katy, TX has a "China town" with its Asian population of 867 (2019 US Census). \
San Antonio does, however, have some smaller grocery stores scattered about the city but no mainstream Asian stores such as H-Mart or 99-Ranch. A medium-sized Asia Super Market opened on the East Side of town, but its debut was met with the COVID pandemic. Most seekers of East Asian cuisine and ingredients have to drive hours to Austin or Houston, TX. My goal for this project is to find an ideal part of San Antonio to open/build a large Asian market in hopes to nucleate a "China town" and increase the cultural diversity of the city. This will probably involve comparing Zip codes' Asian populations and their number of Asian restaurants. Beneficiaries of this research will be the Asian communities and those who appreciate Asian culture and cuisine.

#### An Asian supermarket probably needs to be built where many Asian people are. We can get an idea of the distribution of Asian people in San Antonio from the US census. 

But first let's import the necessary packages...

In [177]:
from bs4 import BeautifulSoup
from urllib.request import urlopen
import pandas as pd
import requests
import numpy as np
import matplotlib.cm as cm
import matplotlib.colors as colors
from sklearn.cluster import KMeans
import folium 
import ssl

### Now, let's get the ZIP code (ZCTA5) demographics for Texas (downloaded from data.census.gov).

In [178]:
demo_df = pd.read_csv('zip_code_demo_2019.csv', skiprows =1)
demo_df.head()

id Geographic Area Name  Estimate!!Total:  \
0  8600000US75001          ZCTA5 75001             14992   
1  8600000US75002          ZCTA5 75002             71253   
2  8600000US75006          ZCTA5 75006             51642   
3  8600000US75007          ZCTA5 75007             55500   
4  8600000US75009          ZCTA5 75009             14089   

  Margin of Error!!Total:  Estimate!!Total:!!White alone  \
0                     219                           8211   
1                    1424                          52482   
2                     547                          40058   
3                    1303                          38046   
4                    1262                          12559   

   Margin of Error!!Total:!!White alone  \
0                                   605   
1                                  1775   
2                                   993   
3                                  1339   
4                                  1691   

   Estimate!!Total:!!Black or African American alone  \
0                                               2194   
1                                               7566   
2                                               4590   
3                                               5078   
4                                               1006   

   Margin of Error!!Total:!!Black or African American alone  \
0                                                381          
1                                                954          
2                                                597          
3                                                775          
4                                                871          

   Estimate!!Total:!!American Indian and Alaska Native alone  \
0                                                115           
1                                                299           
2                                                 80           
3                                                279           
4                                                 37           

   Margin of Error!!Total:!!American Indian and Alaska Native alone  ...  \
0                                                 92                 ...   
1                                                153                 ...   
2                                                 54                 ...   
3                                                145                 ...   
4                                                 40                 ...   

   Estimate!!Total:!!Native Hawaiian and Other Pacific Islander alone  \
0                                                  0                    
1                                                  0                    
2                                                 44                    
3                                                 90                    
4                                                  0                    

   Margin of Error!!Total:!!Native Hawaiian and Other Pacific Islander alone  \
0                                                 21                           
1                                                 32                           
2                                                 46                           
3                                                104                           
4                                                 21                           

   Estimate!!Total:!!Some other race alone  \
0                                     1800   
1                                     1308   
2                                     1955   
3                                     1574   
4                                      191   

   Margin of Error!!Total:!!Some other race alone  \
0                                             603   
1                                             598   
2                                             571   
3                                             506   
4             

#### What are the names of the columns?

In [179]:
demo_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1935 entries, 0 to 1934
Data columns (total 22 columns):
 #   Column                                                                                                     Non-Null Count  Dtype 
---  ------                                                                                                     --------------  ----- 
 0   id                                                                                                         1935 non-null   object
 1   Geographic Area Name                                                                                       1935 non-null   object
 2   Estimate!!Total:                                                                                           1935 non-null   int64 
 3   Margin of Error!!Total:                                                                                    1935 non-null   object
 4   Estimate!!Total:!!White alone                                                 

In [180]:
#Change ZCTA5 ##### to ##### Zipcodes
demo_df['Geographic Area Name'] = demo_df['Geographic Area Name'].str.replace("ZCTA5 ","")
#Change column names
demo_df = demo_df.rename(columns = {'Margin of Error!!Total:!!Asian alone':'Est_Asian', 'Geographic Area Name':'Zip_Code', 
                          'Estimate!!Total:':'Est_Total','Estimate!!Total:!!White alone':'Est_White','Estimate!!Total:!!Black or African American alone':'Est_Black', 
                          'Estimate!!Total:!!American Indian and Alaska Native alone':'Est_Ind_Alaskan_Native','Estimate!!Total:!!Native Hawaiian and Other Pacific Islander alone':'Est_Hawaiian_PI',
                         'Estimate!!Total:!!Some other race alone':'Est_other', 'Estimate!!Total:!!Two or more races:': 'Est_Multiracial',
                         'Estimate!!Total:!!Two or more races:!!Two races including Some other race':'Est_two_other','Estimate!!Total:!!Two or more races:!!Two races excluding Some other race, and three or more races':'Est_two_other_three'})
demo_df.head()

id Zip_Code  Est_Total Margin of Error!!Total:  Est_White  \
0  8600000US75001    75001      14992                     219       8211   
1  8600000US75002    75002      71253                    1424      52482   
2  8600000US75006    75006      51642                     547      40058   
3  8600000US75007    75007      55500                    1303      38046   
4  8600000US75009    75009      14089                    1262      12559   

   Margin of Error!!Total:!!White alone  Est_Black  \
0                                   605       2194   
1                                  1775       7566   
2                                   993       4590   
3                                  1339       5078   
4                                  1691       1006   

   Margin of Error!!Total:!!Black or African American alone  \
0                                                381          
1                                                954          
2                                                597          
3                                                775          
4                                                871          

   Est_Ind_Alaskan_Native  \
0                     115   
1                     299   
2                      80   
3                     279   
4                      37   

   Margin of Error!!Total:!!American Indian and Alaska Native alone  ...  \
0                                                 92                 ...   
1                                                153                 ...   
2                                                 54                 ...   
3                                                145                 ...   
4                                                 40                 ...   

   Est_Hawaiian_PI  \
0                0   
1                0   
2               44   
3               90   
4                0   

   Margin of Error!!Total:!!Native Hawaiian and Other Pacific Islander alone  \
0                                                 21                           
1                                                 32                           
2                                                 46                           
3                                                104                           
4                                                 21                           

   Est_other  Margin of Error!!Total:!!Some other race alone  Est_Multiracial  \
0       1800                                             603             1177   
1       1308                                             598             2396   
2       1955                                             571             1060   
3       1574                                             506             1516   
4        191                                             163              250   

   Margin of Error!!Total:!!Two or more races:  Est_two_other  \
0                                          276            183   
1                                          649            261   
2                                          310            169   
3                                          353            359   
4                                          161             33   

   Margin of Error!!Total:!!Two or more races:!!Two races including Some other race  \
0                                                104                                  
1                                                149                                  
2                                                 87                                  
3                                                153                                  
4                                                 69                                  

   Est_two_other_three  \
0                  994   
1                 2135   
2                  891   
3                 1157   
4                  217   

   Margin of Error!!Total:!!Two or more races:!!Tw

#### Now let's get the coordinates for the zip codes for San Antonio. You can get this information (for the USA) from http://download.geonames.org/export/zip/.

In [181]:
us_zip = pd.read_csv('US.txt', delimiter = '\t', header = None )
us_zip.rename(columns = {0:'Country', 1:'ZIP Code', 2:'Region Name', 3:'State', 4:'State Abbreviation', 5:'County', 9:'Latitude', 10:'Longitude' }, inplace = True)
us_zip.drop(['Country', 6, 7, 8, 11], axis = 1, inplace = True)
#filter for ZIP codes in San Antonio, Texas
tx_zip = us_zip.loc[(us_zip['Region Name'] == 'San Antonio') & (us_zip['State'] == 'Texas')]
tx_zip.head()

ZIP Code  Region Name  State State Abbreviation County  Latitude  \
33958     78201  San Antonio  Texas                 TX  Bexar   29.4711   
33959     78202  San Antonio  Texas                 TX  Bexar   29.4275   
33960     78203  San Antonio  Texas                 TX  Bexar   29.4148   
33961     78204  San Antonio  Texas                 TX  Bexar   29.4059   
33962     78205  San Antonio  Texas                 TX  Bexar   29.4237   

       Longitude  
33958   -98.5356  
33959   -98.4601  
33960   -98.4601  
33961   -98.5078  
33962   -98.4925

#### Now let's merge the SA zip data with the locational data...

In [182]:
tx_zip['ZIP Code'] = tx_zip['ZIP Code'].astype(str)
df_combined = demo_df.merge(tx_zip, how = 'outer', left_on = 'Zip_Code', right_on = 'ZIP Code')
#drop Zip codes not in San Antonio (those with null values)
#df_combined.dropna(subset = ['Latitude'], inplace = True)
df_combined.reset_index(drop=True)
df_combined.head()

/usr/local/Cellar/jupyterlab/3.0.16_1/libexec/lib/python3.9/site-packages/pandas/core/frame.py:3607: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



id Zip_Code  Est_Total Margin of Error!!Total:  Est_White  \
0  8600000US75001    75001    14992.0                     219     8211.0   
1  8600000US75002    75002    71253.0                    1424    52482.0   
2  8600000US75006    75006    51642.0                     547    40058.0   
3  8600000US75007    75007    55500.0                    1303    38046.0   
4  8600000US75009    75009    14089.0                    1262    12559.0   

   Margin of Error!!Total:!!White alone  Est_Black  \
0                                 605.0     2194.0   
1                                1775.0     7566.0   
2                                 993.0     4590.0   
3                                1339.0     5078.0   
4                                1691.0     1006.0   

   Margin of Error!!Total:!!Black or African American alone  \
0                                              381.0          
1                                              954.0          
2                                              597.0          
3                                              775.0          
4                                              871.0          

   Est_Ind_Alaskan_Native  \
0                   115.0   
1                   299.0   
2                    80.0   
3                   279.0   
4                    37.0   

   Margin of Error!!Total:!!American Indian and Alaska Native alone  ...  \
0                                               92.0                 ...   
1                                              153.0                 ...   
2                                               54.0                 ...   
3                                              145.0                 ...   
4                                               40.0                 ...   

   Margin of Error!!Total:!!Two or more races:!!Two races including Some other race  \
0                                              104.0                                  
1                                              149.0                                  
2                                               87.0                                  
3                                              153.0                                  
4                                               69.0                                  

   Est_two_other_three  \
0                994.0   
1               2135.0   
2                891.0   
3               1157.0   
4                217.0   

   Margin of Error!!Total:!!Two or more races:!!Two races excluding Some other race, and three or more races  \
0                                              259.0                                                           
1                                              635.0                                                           
2                                              295.0                                                           
3                                              331.0                                                           
4                                              154.0                                                           

   ZIP Code  Region Name  State  State Abbreviation  County  Latitude  \
0       NaN          NaN    NaN                 NaN     NaN       NaN   
1       NaN          NaN    NaN                 NaN     NaN       NaN   
2       NaN          NaN    NaN                 NaN     NaN       NaN   
3       NaN          NaN    NaN                 NaN     NaN       NaN   
4       NaN          NaN    NaN                 NaN     NaN       NaN   

   Longitude  
0        NaN  
1        NaN  
2        NaN  
3        NaN  
4        NaN  

[5 rows x 29 columns]

In [183]:
df_combined.dropna(subset = ['Latitude'], inplace = True)
df_combined.shape

(83, 29)

29 columns is a lot to handle.
#### Let's make a df that has the Asian population, latitude, longitude and Zip code


In [184]:
df_azn = df_combined[['ZIP Code', 'Est_Asian', 'Est_Hawaiian_PI', 'Est_Total','Latitude', 'Longitude']].reset_index(drop = True)
df_azn.head()

ZIP Code  Est_Asian  Est_Hawaiian_PI  Est_Total  Latitude  Longitude
0    78201      147.0              0.0    48334.0   29.4711   -98.5356
1    78202       51.0              5.0    11453.0   29.4275   -98.4601
2    78203       30.0              0.0     6180.0   29.4148   -98.4601
3    78204       54.0              0.0    12231.0   29.4059   -98.5078
4    78205       14.0              0.0     1441.0   29.4237   -98.4925

Drop any null values in the population (some ZIP codes are post office boxes). Calculate Asian American Pacific Islander pop.

In [185]:
df_azn.dropna(subset = ['Est_Asian'], inplace = True)
df_azn['Est_AAPI'] = df_azn['Est_Asian']+df_azn['Est_Hawaiian_PI']
df_azn.sort_values('Est_AAPI', ascending = False)

ZIP Code  Est_Asian  Est_Hawaiian_PI  Est_Total  Latitude  Longitude  \
36    78240     1487.0            140.0    60927.0   29.5189   -98.6006   
52    78258      856.0              0.0    44747.0   29.6562   -98.4967   
44    78250      676.0            160.0    61669.0   29.5054   -98.6688   
48    78254      595.0            201.0    59107.0   29.5551   -98.7442   
45    78251      719.0             47.0    62186.0   29.4597   -98.6555   
43    78249      609.0             59.0    61639.0   29.5612   -98.6117   
40    78245      424.0            226.0    76518.0   29.4189   -98.6895   
28    78230      631.0              0.0    45021.0   29.5407   -98.5521   
41    78247      541.0             44.0    53892.0   29.5855   -98.4071   
19    78221      491.0             31.0    41185.0   29.3309   -98.5054   
53    78259      471.0             47.0    26338.0   29.6283   -98.4445   
35    78239      322.0            181.0    27431.0   29.5182   -98.3627   
34    78238      401.0             81.0    25660.0   29.4510   -98.6169   
16    78218      467.0              3.0    37281.0   29.4969   -98.4032   
11    78213      415.0             55.0    44813.0   29.5234   -98.5273   
27    78229      417.0             47.0    34125.0   29.5042   -98.5697   
14    78216      417.0             17.0    43554.0   29.5334   -98.4975   
15    78217      392.0             41.0    34159.0   29.5395   -98.4194   
31    78233      362.0             32.0    50103.0   29.5563   -98.3643   
50    78256      333.0             47.0    11238.0   29.6169   -98.6252   
47    78253      356.0              0.0    43770.0   29.4599   -98.7479   
54    78260      311.0             27.0    26301.0   29.7026   -98.4759   
49    78255      328.0              0.0    14825.0   29.6701   -98.6873   
30    78232      302.0              0.0    39072.0   29.5863   -98.4769   
46    78252      232.0             64.0    12009.0   29.3460   -98.6464   
22    78224      284.0              0.0    22851.0   29.3374   -98.5393   
20    78222      267.0              7.0    25227.0   29.3831   -98.3960   
25    78227      246.0             26.0    51394.0   29.4027   -98.6433   
7     78209      252.0             17.0    42653.0   29.4821   -98.4554   
42    78248      167.0             81.0    14345.0   29.5894   -98.5201   
55    78261      220.0             26.0    16749.0   29.7055   -98.4191   
58    78266      210.0             14.0     6913.0   29.6643   -98.3118   
21    78223      224.0              0.0    56055.0   29.3579   -98.4356   
51    78257      222.0              0.0     7538.0   29.6495   -98.6137   
17    78219      220.0              0.0    17239.0   29.4492   -98.3446   
39    78244      205.0             14.0    30981.0   29.4793   -98.3476   
26    78228      192.0             15.0    62359.0   29.4589   -98.5699   
10    78212      131.0             37.0    28865.0   29.4388   -98.4935   
18    78220      165.0              0.0    16700.0   29.4106   -98.4128   
33    78237      150.0              0.0    39105.0   29.4208   -98.5645   
0     78201      147.0              0.0    48334.0   29.4711   -98.5356   
37    78242      131.0             10.0    34893.0   29.3509   -98.6109   
5     78207      118.0              0.0    55484.0   29.4229   -98.5260   
8     78210       84.0             26.0    37683.0   29.3977   -98.4658   
29    78231       88.0              0.0     9001.0   29.5714   -98.5414   
9     78211       68.0             19.0    32769.0   29.3499   -98.5638   
38    78243       58.0             24.0      408.0   29.4375   -98.4616   
12    78214       65.0              0.0    27558.0   29.3641   -98.4924   
13    78215       61.0              0.0     2915.0   29.4413   -98.4793   
56    78263       14.0             43.0     4440.0   29.3614   -98.3174   
1     78202       51.0              5.0    11453.0   29.4275   -98.4601   
57    78264       55.0              0.0    10623.0   29.1733   -98.4723   
3     78204       54.0      

#### 78240 seems to have the largest number of Asians in the city. Hawaiian and Pacific Islander can also be included in the totals (AAPI or Asian American Pacific Islander). What about percent of total pop?

In [186]:
#percent AAPI
df_azn['Percent Asian'] = (df_azn['Est_Asian']+df_azn['Est_Hawaiian_PI'])/df_azn['Est_Total']
#AAPI population for the ZIP codes

df_azn.sort_values('Percent Asian', ascending = False).head()

ZIP Code  Est_Asian  Est_Hawaiian_PI  Est_Total  Latitude  Longitude  \
38    78243       58.0             24.0      408.0   29.4375   -98.4616   
50    78256      333.0             47.0    11238.0   29.6169   -98.6252   
58    78266      210.0             14.0     6913.0   29.6643   -98.3118   
51    78257      222.0              0.0     7538.0   29.6495   -98.6137   
36    78240     1487.0            140.0    60927.0   29.5189   -98.6006   

    Est_AAPI  Percent Asian  
38      82.0       0.200980  
50     380.0       0.033814  
58     224.0       0.032403  
51     222.0       0.029451  
36    1627.0       0.026704

 #### 78240 is still in the top 5 in terms of Asians as percent of the total Zip code population, 
 but the other 4 Zip codes with higher percentages have less Asian people combined than 78240. Interestingly, 78243 has a large percentage of Asian and Pacific Islanders as well. 78243 is part of Lackland Airforce base, and 78240 is adjacent to the Medical Center.

#### In terms of customer base, 78240 is a good zip code. The number of Asian restaurants in each zip code may be another good metric.
First, let's see what kind of venues are in each zip code and cluster the zip codes.

In [30]:
CLIENT_ID = 'GRQDMKG5QXBA34GNMLFNMKLR401CT113G5GY253GY1PNUWVK' # your Foursquare ID
CLIENT_SECRET = 'LZTGZJ5SDZHMAV22XHGOG3HJEFDAAXCE3H5CXEOORJDT11AS' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version
LIMIT = 100 # A default Foursquare API limit value

Function for obtaining venues:

In [31]:
def getNearbyVenues(names, latitudes, longitudes, radius=1000):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Zip Code', 
                  'Zip Code Latitude', 
                  'Zip Code Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

Get data from Foursquare on Venues in San Antonio

In [32]:
sa_venues = getNearbyVenues(names=df_azn['ZIP Code'],
                                   latitudes=df_azn['Latitude'],
                                   longitudes=df_azn['Longitude'])

78201
78202
78203
78204
78205
78207
78208
78209
78210
78211
78212
78213
78214
78215
78216
78217
78218
78219
78220
78221
78222
78223
78224
78225
78226
78227
78228
78229
78230
78231
78232
78233
78235
78237
78238
78239
78240
78242
78243
78244
78245
78247
78248
78249
78250
78251
78252
78253
78254
78255
78256
78257
78258
78259
78260
78261
78263
78264
78266


In [33]:
sa_venues.head()

Zip Code  Zip Code Latitude  Zip Code Longitude                   Venue  \
0    78201            29.4711            -98.5356     Original Donut Shop   
1    78201            29.4711            -98.5356        Restaurant Depot   
2    78201            29.4711            -98.5356               Walgreens   
3    78201            29.4711            -98.5356  De Wese's Tip Top Cafe   
4    78201            29.4711            -98.5356             Tink-A-Tako   

   Venue Latitude  Venue Longitude        Venue Category  
0       29.472703       -98.534598            Donut Shop  
1       29.473163       -98.535505  Kitchen Supply Store  
2       29.473850       -98.534403              Pharmacy  
3       29.470214       -98.531357                  Café  
4       29.477054       -98.539218    Mexican Restaurant

How many venue types?

In [34]:
print('There are {} uniques categories.'.format(len(sa_venues['Venue Category'].unique())))

There are 221 uniques categories.


### Next we'll see what kind of venues are in each ZIP code

In [35]:
# one hot encoding
sa_onehot = pd.get_dummies(sa_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
sa_onehot['Zip Code'] = sa_venues['Zip Code'] 

# move neighborhood column to the first column
sa_onehot = sa_onehot[ ['Zip Code'] + [ col for col in sa_onehot.columns if col != 'Zip Code' ] ]

sa_onehot.head()

Zip Code  Accessories Store  American Restaurant  Arcade  Art Gallery  \
0    78201                  0                    0       0            0   
1    78201                  0                    0       0            0   
2    78201                  0                    0       0            0   
3    78201                  0                    0       0            0   
4    78201                  0                    0       0            0   

   Art Museum  Arts & Crafts Store  Asian Restaurant  Athletics & Sports  \
0           0                    0                 0                   0   
1           0                    0                 0                   0   
2           0                    0                 0                   0   
3           0                    0                 0                   0   
4           0                    0                 0                   0   

   Auto Garage  ...  Vegetarian / Vegan Restaurant  Video Game Store  \
0            0  ...                              0                 0   
1            0  ...                              0                 0   
2            0  ...                              0                 0   
3            0  ...                              0                 0   
4            0  ...                              0                 0   

   Video Store  Vietnamese Restaurant  Warehouse Store  Waterfront  Wine Bar  \
0            0                      0                0           0         0   
1            0                      0                0           0         0   
2            0                      0                0           0         0   
3            0                      0                0           0         0   
4            0                      0                0           0         0   

   Wine Shop  Wings Joint  Yoga Studio  
0          0            0            0  
1          0            0            0  
2          0            0            0  
3          0            0            0  
4          0            0            0  

[5 rows x 222 columns]

How many venues did we obtain?

In [36]:
sa_onehot.shape

(1161, 222)

In [126]:
#count number of each venue type per zip code
sa_sum = sa_onehot.groupby('Zip Code').sum().reset_index()
sa_sum

Zip Code  Accessories Store  American Restaurant  Arcade  Art Gallery  \
0     78201                  0                    0       0            0   
1     78202                  0                    0       0            0   
2     78203                  0                    0       0            0   
3     78204                  0                    0       0            0   
4     78205                  0                    2       0            0   
5     78207                  0                    1       0            0   
6     78208                  0                    0       0            0   
7     78209                  0                    1       0            1   
8     78210                  0                    0       0            0   
9     78211                  0                    0       0            0   
10    78212                  0                    2       0            3   
11    78213                  0                    0       0            0   
12    78214                  0                    0       1            0   
13    78215                  0                    2       0            1   
14    78216                  0                    0       1            0   
15    78217                  0                    0       0            0   
16    78218                  0                    0       0            0   
17    78219                  0                    0       0            0   
18    78220                  0                    0       0            0   
19    78221                  0                    0       0            0   
20    78222                  0                    0       0            0   
21    78223                  0                    1       0            0   
22    78224                  0                    0       0            0   
23    78225                  0                    1       0            0   
24    78226                  1                    0       0            0   
25    78227                  0                    0       0            0   
26    78228                  0                    0       0            0   
27    78229                  0                    0       0            0   
28    78230                  0                    0       0            0   
29    78231                  0                    0       0            0   
30    78232                  0                    0       0            0   
31    78233                  0                    0       0            0   
32    78235                  0                    1       0            0   
33    78237                  0                    0       0            0   
34    78238                  0                    1       0            0   
35    78239                  0                    0       0            0   
36    78240                  0                    1       0            0   
37    78242                  0                    1       0            0   
38    78243                  0                    0       0            0   
39    78244                  0                    0       0            0   
40    78245                  0                    0       0            0   
41    78247                  0                    0       0            0   
42    78248                  0                    1       1            0   
43    78249                  0                    0       0            0   
44    78250                  0                    0       0            0   
45    78251                  0                    0       0            0   
46    78252                  0                    0       0            0   
47    78253                  0                    0       0            0   
48    78255                  0                    0       0            0   
49    78256                  0                    0       0            0   
50    78257                  0                    0       0            0   
51    78258                  0                    0

In [127]:
#group the venue types by ZIP Code
sa_grouped = sa_onehot.groupby('Zip Code').mean().reset_index()
sa_grouped

Zip Code  Accessories Store  American Restaurant    Arcade  Art Gallery  \
0     78201           0.000000             0.000000  0.000000     0.000000   
1     78202           0.000000             0.000000  0.000000     0.000000   
2     78203           0.000000             0.000000  0.000000     0.000000   
3     78204           0.000000             0.000000  0.000000     0.000000   
4     78205           0.000000             0.020000  0.000000     0.000000   
5     78207           0.000000             0.043478  0.000000     0.000000   
6     78208           0.000000             0.000000  0.000000     0.000000   
7     78209           0.000000             0.027027  0.000000     0.027027   
8     78210           0.000000             0.000000  0.000000     0.000000   
9     78211           0.000000             0.000000  0.000000     0.000000   
10    78212           0.000000             0.032787  0.000000     0.049180   
11    78213           0.000000             0.000000  0.000000     0.000000   
12    78214           0.000000             0.000000  0.055556     0.000000   
13    78215           0.000000             0.025641  0.000000     0.012821   
14    78216           0.000000             0.000000  0.043478     0.000000   
15    78217           0.000000             0.000000  0.000000     0.000000   
16    78218           0.000000             0.000000  0.000000     0.000000   
17    78219           0.000000             0.000000  0.000000     0.000000   
18    78220           0.000000             0.000000  0.000000     0.000000   
19    78221           0.000000             0.000000  0.000000     0.000000   
20    78222           0.000000             0.000000  0.000000     0.000000   
21    78223           0.000000             0.017857  0.000000     0.000000   
22    78224           0.000000             0.000000  0.000000     0.000000   
23    78225           0.000000             0.038462  0.000000     0.000000   
24    78226           0.034483             0.000000  0.000000     0.000000   
25    78227           0.000000             0.000000  0.000000     0.000000   
26    78228           0.000000             0.000000  0.000000     0.000000   
27    78229           0.000000             0.000000  0.000000     0.000000   
28    78230           0.000000             0.000000  0.000000     0.000000   
29    78231           0.000000             0.000000  0.000000     0.000000   
30    78232           0.000000             0.000000  0.000000     0.000000   
31    78233           0.000000             0.000000  0.000000     0.000000   
32    78235           0.000000             0.021277  0.000000     0.000000   
33    78237           0.000000             0.000000  0.000000     0.000000   
34    78238           0.000000             0.125000  0.000000     0.000000   
35    78239           0.000000             0.000000  0.000000     0.000000   
36    78240           0.000000             0.018868  0.000000     0.000000   
37    78242           0.000000             0.071429  0.000000     0.000000   
38    78243           0.000000             0.000000  0.000000     0.000000   
39    78244           0.000000             0.000000  0.000000     0.000000   
40    78245           0.000000             0.000000  0.000000     0.000000   
41    78247           0.000000             0.000000  0.000000     0.000000   
42    78248           0.000000             0.037037  0.037037     0.000000   
43    78249           0.000000             0.000000  0.000000     0.000000   
44    78250           0.000000             0.000000  0.000000     0.000000   
45    78251           0.000000             0.000000  0.000000     0.000000   
46    78252           0.000000             0.000000  0.000000     0.000000   
47    78253           0.000000             0.000000  0.000000     0.000000   
48    78255           0.000000             0.000000  0.000000     0.000000   
49    78256           0.000000             0.000000  0.000000     0.000000   
50    78257           0.0

In [39]:
len(sa_grouped.columns)

222

#### Now to see if any stand out for Asian Restaurants...

In [40]:
num_top_venues = 5

for hood in sa_grouped['Zip Code']:
    print("----"+hood+"----")
    temp = sa_grouped[sa_grouped['Zip Code'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----78201----
                  venue  freq
0    Mexican Restaurant  0.22
1     Convenience Store  0.07
2  Kitchen Supply Store  0.04
3             Pet Store  0.04
4             Pawn Shop  0.04


----78202----
                venue  freq
0  Mexican Restaurant  0.12
1           BBQ Joint  0.06
2   Mobile Phone Shop  0.06
3       Grocery Store  0.06
4                Café  0.06


----78203----
                             venue  freq
0               Mexican Restaurant  0.29
1              Fried Chicken Joint  0.14
2  Southern / Soul Food Restaurant  0.14
3                     Home Service  0.14
4                             Park  0.14


----78204----
                    venue  freq
0      Mexican Restaurant  0.28
1                     Bar  0.08
2             Pizza Place  0.08
3  Gluten-free Restaurant  0.04
4                 Brewery  0.04


----78205----
                venue  freq
0               Hotel  0.15
1  Mexican Restaurant  0.06
2             Theater  0.05
3  Seafood Restaurant  0

 So in terms of frequency between venues, 78222 and 78229 have a few Chinese restaurants, and 78259 has a few Thai Restaurants. 

#### Let's try explicitly searching with Foursquare for Asian restaurants.

In [108]:
def getNearbyAsianRest(names, latitudes, longitudes, radius=1200):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL category id is for 'Asian Restaurants'
        url = 'https://api.foursquare.com/v2/venues/search?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}&categoryId=4bf58dd8d48988d142941735'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['venues']
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['name'], 
            v['location']['lat'], 
            v['location']['lng'],  
            v['categories'][0]['name']
    ) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Zip Code', 
                  'Zip Code Latitude', 
                  'Zip Code Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category',
                            ]
    
    return(nearby_venues)

In [109]:
#call the modified function
sa_rest = getNearbyAsianRest(names=df_azn['ZIP Code'],
                                   latitudes=df_azn['Latitude'],
                                   longitudes=df_azn['Longitude'])

78201
78202
78203
78204
78205
78207
78208
78209
78210
78211
78212
78213
78214
78215
78216
78217
78218
78219
78220
78221
78222
78223
78224
78225
78226
78227
78228
78229
78230
78231
78232
78233
78235
78237
78238
78239
78240
78242
78243
78244
78245
78247
78248
78249
78250
78251
78252
78253
78254
78255
78256
78257
78258
78259
78260
78261
78263
78264
78266


### How many Asian Restaurants are their in each ZIP code?

In [110]:
sa_rest_count= sa_rest.groupby('Zip Code').count().sort_values('Venue', ascending = False)
sa_rest_count['Venue']

Zip Code
78229    18
78205    18
78212    13
78218    11
78251     9
78240     9
78227     9
78215     8
78217     7
78216     6
78213     6
78232     4
78249     4
78214     4
78259     4
78204     4
78223     4
78235     3
78237     3
78243     3
78208     3
78202     2
78222     2
78230     2
78248     2
78233     2
78203     2
78201     2
78250     1
78257     1
78244     1
78226     1
78239     1
78238     1
78231     1
78225     1
78224     1
78220     1
78209     1
78207     1
78260     1
Name: Venue, dtype: int64

 
 Looks like from the last analysis, 78229 has the most number of Asian restaurants (within 2 km). 78240 is in 12th place.

#### How close are all of these Zip codes to each other?

In [111]:
df_mapping = sa_rest_count.merge(df_azn, how = 'outer', left_on = 'Zip Code', right_on = 'ZIP Code')
df_mapping.drop(axis = 1, labels = ['Zip Code Latitude', 'Zip Code Longitude', 'Venue Latitude', 'Venue Longitude', 'Venue Category'], inplace =True)
df_mapping

Venue ZIP Code  Est_Asian  Est_Hawaiian_PI  Est_Total  Latitude  \
0    18.0    78229      417.0             47.0    34125.0   29.5042   
1    18.0    78205       14.0              0.0     1441.0   29.4237   
2    13.0    78212      131.0             37.0    28865.0   29.4388   
3    11.0    78218      467.0              3.0    37281.0   29.4969   
4     9.0    78251      719.0             47.0    62186.0   29.4597   
5     9.0    78240     1487.0            140.0    60927.0   29.5189   
6     9.0    78227      246.0             26.0    51394.0   29.4027   
7     8.0    78215       61.0              0.0     2915.0   29.4413   
8     7.0    78217      392.0             41.0    34159.0   29.5395   
9     6.0    78216      417.0             17.0    43554.0   29.5334   
10    6.0    78213      415.0             55.0    44813.0   29.5234   
11    4.0    78232      302.0              0.0    39072.0   29.5863   
12    4.0    78249      609.0             59.0    61639.0   29.5612   
13    4.0    78214       65.0              0.0    27558.0   29.3641   
14    4.0    78259      471.0             47.0    26338.0   29.6283   
15    4.0    78204       54.0              0.0    12231.0   29.4059   
16    4.0    78223      224.0              0.0    56055.0   29.3579   
17    3.0    78235       17.0              0.0     1476.0   29.3494   
18    3.0    78237      150.0              0.0    39105.0   29.4208   
19    3.0    78243       58.0             24.0      408.0   29.4375   
20    3.0    78208       14.0              0.0     4913.0   29.4400   
21    2.0    78202       51.0              5.0    11453.0   29.4275   
22    2.0    78222      267.0              7.0    25227.0   29.3831   
23    2.0    78230      631.0              0.0    45021.0   29.5407   
24    2.0    78248      167.0             81.0    14345.0   29.5894   
25    2.0    78233      362.0             32.0    50103.0   29.5563   
26    2.0    78203       30.0              0.0     6180.0   29.4148   
27    2.0    78201      147.0              0.0    48334.0   29.4711   
28    1.0    78250      676.0            160.0    61669.0   29.5054   
29    1.0    78257      222.0              0.0     7538.0   29.6495   
30    1.0    78244      205.0             14.0    30981.0   29.4793   
31    1.0    78226       19.0              0.0     7906.0   29.3930   
32    1.0    78239      322.0            181.0    27431.0   29.5182   
33    1.0    78238      401.0             81.0    25660.0   29.4510   
34    1.0    78231       88.0              0.0     9001.0   29.5714   
35    1.0    78225       49.0              0.0    14472.0   29.3875   
36    1.0    78224      284.0              0.0    22851.0   29.3374   
37    1.0    78220      165.0              0.0    16700.0   29.4106   
38    1.0    78209      252.0             17.0    42653.0   29.4821   
39    1.0    78207      118.0              0.0    55484.0   29.4229   
40    1.0    78260      311.0             27.0    26301.0   29.7026   
41    NaN    78210       84.0             26.0    37683.0   29.3977   
42    NaN    78211       68.0             19.0    32769.0   29.3499   
43    NaN    78219      220.0              0.0    17239.0   29.4492   
44    NaN    78221      491.0             31.0    41185.0   29.3309   
45    NaN    78228      192.0             15.0    62359.0   29.4589   
46    NaN    78242      131.0             10.0    34893.0   29.3509   
47    NaN    78245      424.0            226.0    76518.0   29.4189   
48    NaN    78247      541.0             44.0    53892.0   29.5855   
49    NaN    78252      232.0             64.0    12009.0   29.3460   
50    NaN    78253      356.0              0.0    43770.0   29.4599   
51    NaN    78254      595.0            201.0    59107.0   29.5551   
52    NaN    78255      328.0              0.0    14825.0   29.6701   
53    NaN    78256      333.0             47.0    11238.0   29.6169   
54    NaN    78258      856.0              0.0    44747.0   29.6562   
55    NaN    78261      220.

In [112]:
#drop nulls
df_mapping.dropna(axis = 0, how = 'any', inplace = True)

In [122]:
# function to map the ZIP codes, putting bubbles sized by number of Asian Restaurants
import folium
sa_coordinates = [df_mapping['Latitude'].mean(), df_mapping['Longitude'].mean()]
m = folium.Map(location=sa_coordinates, tiles="OpenStreetMap", zoom_start=11)
for i in range(0,len(df_mapping)):
   folium.Circle(
      location=[df_mapping.iloc[i]['Latitude'], df_mapping.iloc[i]['Longitude']],
      popup=df_mapping.iloc[i]['ZIP Code'],
      radius=float(df_mapping.iloc[i]['Venue'])*50,
      color='blue',
      fill=False,
      fill_color='blue'
   ).add_to(m)

m

#### Two areas seem to have a higher density of Asian Restaurants: Downtown and the Medical Center to the Northwest.

Let's take a closer look at the downtown ZIP codes...

In [104]:
sa_rest.loc[sa_rest['Zip Code'] == '78212']

Zip Code  Zip Code Latitude  Zip Code Longitude  \
33    78212            29.4388            -98.4935   
34    78212            29.4388            -98.4935   
35    78212            29.4388            -98.4935   
36    78212            29.4388            -98.4935   
37    78212            29.4388            -98.4935   
38    78212            29.4388            -98.4935   
39    78212            29.4388            -98.4935   
40    78212            29.4388            -98.4935   
41    78212            29.4388            -98.4935   
42    78212            29.4388            -98.4935   
43    78212            29.4388            -98.4935   
44    78212            29.4388            -98.4935   
45    78212            29.4388            -98.4935   

                           Venue  Venue Latitude  Venue Longitude  \
33                Saigon Express       29.445055       -98.491520   
34  Pinch Boil House and Bia Bar       29.425906       -98.493674   
35         Best Quality Daughter       29.441435       -98.479240   
36                         Tenko       29.442165       -98.480469   
37                         Shiro       29.436195       -98.481240   
38                Curry Boys Bbq       29.448236       -98.487560   
39  Sze-Chuen Chinese Restaurant       29.426003       -98.493467   
40                Pho Bo Express       29.427437       -98.487520   
41              Teriyaki Kitchen       29.425472       -98.500511   
42                        Botika       29.442526       -98.479975   
43                        Kimura       29.428522       -98.491441   
44                 Far East Cafe       29.436264       -98.501909   
45               Thai Lao Orchid       29.427182       -98.487769   

           Venue Category  
33  Vietnamese Restaurant  
34     Seafood Restaurant  
35     Chinese Restaurant  
36       Ramen Restaurant  
37    Japanese Restaurant  
38       Asian Restaurant  
39     Chinese Restaurant  
40  Vietnamese Restaurant  
41     Chinese Restaurant  
42    Peruvian Restaurant  
43           Noodle House  
44       Asian Restaurant  
45       Asian Restaurant

In [105]:
sa_rest.loc[sa_rest['Zip Code'] == '78205']

Zip Code  Zip Code Latitude  Zip Code Longitude  \
10    78205            29.4237            -98.4925   
11    78205            29.4237            -98.4925   
12    78205            29.4237            -98.4925   
13    78205            29.4237            -98.4925   
14    78205            29.4237            -98.4925   
15    78205            29.4237            -98.4925   
16    78205            29.4237            -98.4925   
17    78205            29.4237            -98.4925   
18    78205            29.4237            -98.4925   
19    78205            29.4237            -98.4925   
20    78205            29.4237            -98.4925   
21    78205            29.4237            -98.4925   
22    78205            29.4237            -98.4925   
23    78205            29.4237            -98.4925   
24    78205            29.4237            -98.4925   
25    78205            29.4237            -98.4925   
26    78205            29.4237            -98.4925   
27    78205            29.4237            -98.4925   

                           Venue  Venue Latitude  Venue Longitude  \
10                   Sushi Zushi       29.424015       -98.491907   
11                    Blue Whale       29.423884       -98.493056   
12  Pinch Boil House and Bia Bar       29.425906       -98.493674   
13                        Bamboo       29.413464       -98.499669   
14  Sze-Chuen Chinese Restaurant       29.426003       -98.493467   
15                    Thai Lucky       29.421184       -98.490475   
16                 Side Wok Cafe       29.423517       -98.484732   
17                  Sansei Sushi       29.421921       -98.484116   
18                       Hot Joy       29.413809       -98.490913   
19                Pho Bo Express       29.427437       -98.487520   
20                  Little Tokyo       29.423567       -98.484664   
21              Golden Star Cafe       29.426116       -98.503437   
22              Teriyaki Kitchen       29.425472       -98.500511   
23                      Bar Rojo       29.422222       -98.484600   
24                         Tiwan       29.425104       -98.493118   
25    April's Chinese Restaurant       29.413613       -98.504163   
26                        Kimura       29.428522       -98.491441   
27               Thai Lao Orchid       29.427182       -98.487769   

           Venue Category  
10       Sushi Restaurant  
11    Japanese Restaurant  
12     Seafood Restaurant  
13     Chinese Restaurant  
14     Chinese Restaurant  
15        Thai Restaurant  
16       Asian Restaurant  
17       Sushi Restaurant  
18       Asian Restaurant  
19  Vietnamese Restaurant  
20       Asian Restaurant  
21     Chinese Restaurant  
22     Chinese Restaurant  
23              Hotel Bar  
24       Asian Restaurant  
25     Chinese Restaurant  
26           Noodle House  
27       Asian Restaurant

Seems like a few restaurants are shared between 78212 and 78205 (Kimura, Teriyaki Kitchen, Sze-chuan, Pinch Boil House) which are fairly small ZIP codes. This is one of the limits of this analysis - ZIP codes are not all equal area.

#### What if we add circles to the map representing AAPI population for the ZIP codes?

In [123]:
for i in range(0,len(df_mapping)):
   folium.Circle(
      location=[df_mapping.iloc[i]['Latitude'], df_mapping.iloc[i]['Longitude']],
      popup=df_mapping.iloc[i]['ZIP Code'],
      radius=float(df_mapping.iloc[i]['Est_AAPI']),
      color='red',
      fill=False,
      fill_color='red'
   ).add_to(m)
m

With the red circles representing AAPI population, the Northwest side of San Antonio has more AAPI people (likely customers) than Downtown San Antonio. Northwest San Antonio appears to be the best location for a large Asian supermarket, with 78240 having the most AAPI residents and 78229 having the most Asian restaurants.

Let's see if we can find similarities between the ZIP codes in terms of the restaurants collected...

In [130]:
# one hot encoding
rest_onehot = pd.get_dummies(sa_rest[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
rest_onehot['Zip Code'] = sa_rest['Zip Code'] 

# move neighborhood column to the first column
rest_onehot = rest_onehot[ ['Zip Code'] + [ col for col in rest_onehot.columns if col != 'Zip Code' ] ]

rest_onehot.head()

Zip Code  Asian Restaurant  Cantonese Restaurant  Chinese Restaurant  \
0    78201                 0                     0                   1   
1    78201                 0                     0                   0   
2    78202                 0                     0                   1   
3    78202                 0                     0                   0   
4    78203                 0                     0                   0   

   Dim Sum Restaurant  Filipino Restaurant  Food Service  Food Truck  \
0                   0                    0             0           0   
1                   0                    0             0           0   
2                   0                    0             0           0   
3                   0                    0             0           0   
4                   0                    0             0           0   

   Hotel Bar  Japanese Restaurant  Korean Restaurant  Noodle House  \
0          0                    0                  0             0   
1          0                    0                  0             0   
2          0                    0                  0             0   
3          0                    0                  1             0   
4          0                    0                  0             0   

   Peruvian Restaurant  Ramen Restaurant  Seafood Restaurant  \
0                    0                 0                   0   
1                    0                 0                   0   
2                    0                 0                   0   
3                    0                 0                   0   
4                    0                 0                   0   

   Sushi Restaurant  Thai Restaurant  Vietnamese Restaurant  
0                 0                0                      0  
1                 0                0                      1  
2                 0                0                      0  
3                 0                0                      0  
4                 0                1                      0

In [132]:
sa_grouped1 = rest_onehot.groupby('Zip Code').mean().reset_index()
sa_grouped1

Zip Code  Asian Restaurant  Cantonese Restaurant  Chinese Restaurant  \
0     78201          0.000000                  0.00            0.500000   
1     78202          0.000000                  0.00            0.500000   
2     78203          0.000000                  0.00            0.500000   
3     78204          0.000000                  0.00            0.500000   
4     78205          0.277778                  0.00            0.277778   
5     78207          0.000000                  0.00            1.000000   
6     78208          0.333333                  0.00            0.333333   
7     78209          0.000000                  0.00            0.000000   
8     78212          0.230769                  0.00            0.230769   
9     78213          0.166667                  0.00            0.500000   
10    78214          0.000000                  0.00            1.000000   
11    78215          0.125000                  0.00            0.125000   
12    78216          0.166667                  0.00            0.500000   
13    78217          0.000000                  0.00            0.428571   
14    78218          0.181818                  0.00            0.000000   
15    78220          0.000000                  0.00            1.000000   
16    78222          0.000000                  0.00            1.000000   
17    78223          0.250000                  0.00            0.750000   
18    78224          1.000000                  0.00            0.000000   
19    78225          0.000000                  0.00            1.000000   
20    78226          0.000000                  0.00            1.000000   
21    78227          0.444444                  0.00            0.333333   
22    78229          0.111111                  0.00            0.333333   
23    78230          0.000000                  0.00            0.500000   
24    78231          0.000000                  0.00            0.000000   
25    78232          0.250000                  0.00            0.250000   
26    78233          0.000000                  0.00            0.500000   
27    78235          0.000000                  0.00            0.666667   
28    78237          0.000000                  0.00            1.000000   
29    78238          1.000000                  0.00            0.000000   
30    78239          0.000000                  0.00            1.000000   
31    78240          0.333333                  0.00            0.333333   
32    78243          0.000000                  0.00            0.666667   
33    78244          0.000000                  0.00            0.000000   
34    78248          0.000000                  0.00            0.500000   
35    78249          0.000000                  0.25            0.000000   
36    78250          0.000000                  0.00            1.000000   
37    78251          0.111111                  0.00            0.222222   
38    78257          0.000000                  0.00            0.000000   
39    78259          0.000000                  0.00            0.000000   
40    78260          0.000000                  0.00            1.000000   

    Dim Sum Restaurant  Filipino Restaurant  Food Service  Food Truck  \
0             0.000000             0.000000      0.000000        0.00   
1             0.000000             0.000000      0.000000        0.00   
2             0.000000             0.000000      0.000000        0.00   
3             0.000000             0.000000      0.000000        0.00   
4             0.000000             0.000000      0.000000        0.00   
5             0.000000             0.000000      0.000000        0.00   
6             0.000000             0.000000      0.000000        0.00   
7             0.000000             0.000000      0.000000        0.00   
8             0.000000             0.000000      0.000000        0.00   
9             0.000000             0.000000      0.000000        0.00   
10            0.000000             0.000000      

Let's take a look at the Asian Restaurants in each Zip code

In [140]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Zip Code']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
zip_venues_sorted = pd.DataFrame(columns=columns)
zip_venues_sorted['Zip Code'] = sa_grouped1['Zip Code']

for ind in np.arange(sa_grouped1.shape[0]):
    zip_venues_sorted.iloc[ind, 1:] = return_most_common_venues(sa_grouped1.iloc[ind, :], num_top_venues)

zip_venues_sorted.head()

Zip Code  1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0    78201  Vietnamese Restaurant    Chinese Restaurant     Korean Restaurant   
1    78202      Korean Restaurant    Chinese Restaurant      Asian Restaurant   
2    78203     Chinese Restaurant       Thai Restaurant      Asian Restaurant   
3    78204     Chinese Restaurant      Sushi Restaurant      Ramen Restaurant   
4    78205       Asian Restaurant    Chinese Restaurant      Sushi Restaurant   

  4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
0       Thai Restaurant      Sushi Restaurant    Seafood Restaurant   
1       Thai Restaurant      Sushi Restaurant    Seafood Restaurant   
2     Korean Restaurant      Sushi Restaurant    Seafood Restaurant   
3      Asian Restaurant     Korean Restaurant       Thai Restaurant   
4             Hotel Bar       Thai Restaurant    Seafood Restaurant   

  7th Most Common Venue 8th Most Common Venue  9th Most Common Venue  \
0      Ramen Restaurant   Peruvian Restaurant           Noodle House   
1      Ramen Restaurant   Peruvian Restaurant           Noodle House   
2      Ramen Restaurant   Peruvian Restaurant           Noodle House   
3    Seafood Restaurant   Peruvian Restaurant           Noodle House   
4          Noodle House   Japanese Restaurant  Vietnamese Restaurant   

  10th Most Common Venue  
0       Asian Restaurant  
1    Japanese Restaurant  
2    Japanese Restaurant  
3    Japanese Restaurant  
4   Cantonese Restaurant

In [134]:
num_top_venues = 5

for hood in sa_grouped1['Zip Code']:
    print("----"+hood+"----")
    temp = sa_grouped1[sa_grouped1['Zip Code'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----78201----
                   venue  freq
0  Vietnamese Restaurant   0.5
1     Chinese Restaurant   0.5
2      Korean Restaurant   0.0
3        Thai Restaurant   0.0
4       Sushi Restaurant   0.0


----78202----
                venue  freq
0   Korean Restaurant   0.5
1  Chinese Restaurant   0.5
2    Asian Restaurant   0.0
3     Thai Restaurant   0.0
4    Sushi Restaurant   0.0


----78203----
                venue  freq
0  Chinese Restaurant   0.5
1     Thai Restaurant   0.5
2    Asian Restaurant   0.0
3   Korean Restaurant   0.0
4    Sushi Restaurant   0.0


----78204----
                venue  freq
0  Chinese Restaurant  0.50
1    Sushi Restaurant  0.25
2    Ramen Restaurant  0.25
3    Asian Restaurant  0.00
4   Korean Restaurant  0.00


----78205----
                venue  freq
0    Asian Restaurant  0.28
1  Chinese Restaurant  0.28
2    Sushi Restaurant  0.11
3           Hotel Bar  0.06
4     Thai Restaurant  0.06


----78207----
                venue  freq
0  Chinese Restauran

Some ZIP codes have no Asian restaurants, some have quite the diversity.

Let's prepare the data for clustering.

In [135]:
#Function for returning the most common venues in descending order
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Now to cluster the ZIP codes...

In [155]:
# set number of clusters
kclusters = 5

rest_grouped_clustering = sa_grouped1.drop('Zip Code', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(rest_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

<ipython-input-155-3fb3188808e7>:4: FutureWarning:

In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only



array([0, 0, 0, 0, 0, 1, 0, 2, 0, 0], dtype=int32)

In [172]:
#sometimes the Cluster Labels column needs to be dropped...
#zip_venues_sorted.drop(axis =1, labels = 'Cluster Labels', inplace = True)

In [169]:
# add clustering labels
zip_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)
sa_merged = df_mapping
# merge sa_grouped with df_mapping (sa data) to add latitude/longitude for each neighborhood
sa_merged = sa_merged.merge(zip_venues_sorted, how = 'outer', left_on = 'ZIP Code', right_on = 'Zip Code')
sa_merged.head() # check the last columns!

Venue ZIP Code  Est_Asian  Est_Hawaiian_PI  Est_Total  Latitude  Longitude  \
0   18.0    78229      417.0             47.0    34125.0   29.5042   -98.5697   
1   18.0    78205       14.0              0.0     1441.0   29.4237   -98.4925   
2   13.0    78212      131.0             37.0    28865.0   29.4388   -98.4935   
3   11.0    78218      467.0              3.0    37281.0   29.4969   -98.4032   
4    9.0    78251      719.0             47.0    62186.0   29.4597   -98.6555   

   Percent Asian  Est_AAPI  Cluster Labels  ...  1st Most Common Venue  \
0       0.013597     464.0               0  ...     Chinese Restaurant   
1       0.009715      14.0               0  ...       Asian Restaurant   
2       0.005820     168.0               0  ...       Asian Restaurant   
3       0.012607     470.0               0  ...      Korean Restaurant   
4       0.012318     766.0               0  ...  Vietnamese Restaurant   

  2nd Most Common Venue  3rd Most Common Venue  4th Most Common Venue  \
0       Thai Restaurant       Asian Restaurant  Vietnamese Restaurant   
1    Chinese Restaurant       Sushi Restaurant              Hotel Bar   
2    Chinese Restaurant  Vietnamese Restaurant           Noodle House   
3      Asian Restaurant  Vietnamese Restaurant           Food Service   
4    Chinese Restaurant        Thai Restaurant    Filipino Restaurant   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0     Korean Restaurant      Sushi Restaurant   Japanese Restaurant   
1       Thai Restaurant    Seafood Restaurant          Noodle House   
2   Peruvian Restaurant      Ramen Restaurant    Seafood Restaurant   
3    Dim Sum Restaurant   Filipino Restaurant            Food Truck   
4      Asian Restaurant    Dim Sum Restaurant          Food Service   

  8th Most Common Venue  9th Most Common Venue 10th Most Common Venue  
0   Filipino Restaurant           Food Service             Food Truck  
1   Japanese Restaurant  Vietnamese Restaurant   Cantonese Restaurant  
2   Japanese Restaurant     Dim Sum Restaurant    Filipino Restaurant  
3             Hotel Bar   Cantonese Restaurant     Chinese Restaurant  
4            Food Truck              Hotel Bar   Cantonese Restaurant  

[5 rows x 21 columns]

Now plot the clusters on the map!

In [171]:
# Latitude of San Antonio? Try mean of the lat/lng table?
latitude = df_mapping['Latitude'].mean()
longitude = df_mapping['Longitude'].mean()
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(sa_merged['Latitude'], sa_merged['Longitude'], sa_merged['ZIP Code'], sa_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster-1)],
        fill=True,
        fill_color=rainbow[int(cluster-1)],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

Interestingly, the North side of San Antonio is fairly similar to Downtown in terms of the kinds of Asian Restaurants, at least according to the K Means fit. This does not affect the conclusion that Northwest San Antonio is a good area for opening a large Asian supermarket. 


In [173]:
#the map we made earlier
#size of blue bubble related to # of Asian restaurants
#size of red bubbles related to # of AAPI residents
m